cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -0,0 +1,640 @@
+// JSON-TO-STACK-OP DFA
+//------------------------------------------------------------------------------
+namespace to_stack_op {
+
+// Type used to represent the target state in the transition table
+using StateT = char;
+
+/**
+ * @brief Definition of the DFA's states
+ */
+enum DFA_STATES : StateT {
+  // The state being active while being outside of a string. When encountering an opening bracket
+  // or curly brace, we push it onto the stack. When encountering a closing bracket or brace, we
+  // pop from the stack.
+  TT_OOS = 0U,
+
+  // The state being active while being within a string (e.g., field name or a string value). We do
+  // not push or pop from the stack while being in this state.
+  TT_STR,
+
+  // The state being active after encountering an escape symbol (e.g., '\'), while being in the
+  // TT_STR state.
+  TT_ESC,
+
+  // Total number of states
+  TT_NUM_STATES
+};
+
+/**
+ * @brief Definition of the symbol groups
+ */
+enum class DFASymbolGroupID : uint32_t {
+  OpenBrace,         ///< Opening brace SG: {
+  OpenBracket,       ///< Opening bracket SG: [
+  CloseBrace,        ///< Closing brace SG: }
+  CloseBracket,      ///< Closing bracket SG: ]
+  Quote,             ///< Quote character SG: "
+  Escape,            ///< Escape character SG: '\'
+  Other,             ///< SG implicitly matching all other characters
+  NUM_SYMBOL_GROUPS  ///< Total number of symbol groups
+};
+
+// The i-th string representing all the characters of a symbol group
+const std::vector<std::string> symbol_groups = {"{", "[", "}", "]", "\"", "\\"};
+
+// Transition table
+const std::vector<std::vector<StateT>> transition_table = {
+  /* IN_STATE         {       [       }       ]       "       \    OTHER */
+  /* TT_OOS    */ {TT_OOS, TT_OOS, TT_OOS, TT_OOS, TT_STR, TT_OOS, TT_OOS},
+  /* TT_STR    */ {TT_STR, TT_STR, TT_STR, TT_STR, TT_OOS, TT_ESC, TT_STR},
+  /* TT_ESC    */ {TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR}};
+
+// Translation table (i.e., for each transition, what are the symbols that we output)
+const std::vector<std::vector<std::vector<char>>> translation_table = {
+  /* IN_STATE        {      [      }      ]     "  \   OTHER */
+  /* TT_OOS    */ {{'{'}, {'['}, {'}'}, {']'}, {'x'}, {'x'}, {'x'}},
+  /* TT_STR    */ {{'x'}, {'x'}, {'x'}, {'x'}, {'x'}, {'x'}, {'x'}},
+  /* TT_ESC    */ {{'x'}, {'x'}, {'x'}, {'x'}, {'x'}, {'x'}, {'x'}}};
+
+// The DFA's starting state
+constexpr auto start_state = TT_OOS;
+}  // namespace to_stack_op
+
+//------------------------------------------------------------------------------
+// JSON TOKENIZER PUSHDOWN AUTOMATON
+//------------------------------------------------------------------------------
+namespace tokenizer_pda {
+
+// Type used to represent the target state in the transition table
+using StateT = char;
+
+/**
+ * @brief Symbol groups for the input alphabet for the pushdown automaton
+ */
+enum SGID : PdaSymbolGroupIdT {
+  /// Opening brace
+  OBC,
```

Comment:
Do we need these to be awkward acronyms, or can they be self-describing names? Also +1 for `enum class` wherever possible. I pulled the capitalization rules from [other libcudf examples of `enum class`](https://github.com/rapidsai/cudf/search?q=%22enum+class%22), which generally use `snake_case` for the enum class name and `ALL_CAPS` for the values.
```suggestion
enum class symbol_group_id : PdaSymbolGroupIdT {
  /// Opening brace
  OPENING_BRACE,
```

Reply:
AFAIK, 3 letter names for enum values are given so that code for the transition and translation tables is readable. Otherwise, I prefer self-describing names too.

FYI, @elstehle If after using long names here, if formatting is messed up due to clang-format, you could wrap the table code lines around `// clang-format off` and `// clang-format on`.

Reply:
It’d also be fine to define a local value like `auto const OBC = symbol_group_id::OPENING_BRACE;` if that helps make the table readable. If these are only used in the table, then maybe it doesn’t matter if the names are readable. I just want to avoid spreading code across multiple files like `if (symbol == symbol_group_id::OBC)` and have to go find what the values mean.

Reply:
Yeah, that looks better. I've fixed the capitalization 👍 

I'm not sure, though, how to accommodate `enum class` for the transition&translation tables without having everything wrapped in `static_cast` hinder readability. The scope of these enums is limited to the definition of the tables. 

Would creating aliases improve things? Any other suggestions? 
```
constexpr auto TT_OOS = static_cast<StateT>(dfa_states::TT_OOS);
constexpr auto TT_STR = static_cast<StateT>(dfa_states::TT_STR);
constexpr auto TT_ESC = static_cast<StateT>(dfa_states::TT_ESC);
```

This is their use:
```
const std::vector<std::vector<StateT>> transition_table = {
  /* IN_STATE         {       [       }       ]       "       \    OTHER */
  /* TT_OOS    */ {TT_OOS, TT_OOS, TT_OOS, TT_OOS, TT_STR, TT_OOS, TT_OOS},
  /* TT_STR    */ {TT_STR, TT_STR, TT_STR, TT_STR, TT_OOS, TT_ESC, TT_STR},
  /* TT_ESC    */ {TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR}};
```

Reply:
I've given self-describing names to anything that would not degrade transition/translation table readability. `enum` versus `enum class` is yet outstanding. Open for suggestions on how to proceed with `enum class` for the lookup tables.

Reply:
Can we define `transition_table` as `const std::vector<std::vector<dfa_states>>` as opposed to vector of `StateT`? Then use aliases for clarity if needed: https://godbolt.org/z/WrKooMxTf

Reply:
> Can we define `transition_table` as `const std::vector<std::vector<dfa_states>>` as opposed to vector of `StateT`? Then use aliases for clarity if needed: https://godbolt.org/z/WrKooMxTf

Yup, agree, I think that's the most reasonable way to accommodate `enum class`es. I've adopted your suggestion. Thanks 👍